versioned/storage/store/src/main/java/org/projectnessie/versioned/storage/versionstore/VersionStoreImpl.java

```diff
@@ -591,20 +594,32 @@ public PaginationIterator<KeyEntry> getKeys(
             ContentKey key = storeKeyToKey(indexElement.key());
             CommitOp commitOp = indexElement.content();
             Content.Type contentType = contentTypeForPayload(commitOp.payload());
+            Content content = null;
 
-            if (withContent) {
-              Content c =
+            UUID contentId = commitOp.contentId();
+            String contentIdString;
+            if (contentId != null) {
+              contentIdString = contentId.toString();
+            } else {
+              // this should only be hit by imported legacy nessie repos
+              content =
                   contentMapping.fetchContent(
-                      requireNonNull(
-                          indexElement.content().value(), "Required value pointer is null"));
-              return KeyEntry.of(buildIdentifiedKey(key, index, c, x -> null), c);
+                      requireNonNull(commitOp.value(), "Required value pointer is null"));
+              contentIdString = content.getId();
             }
-
-            UUID contentId = commitOp.contentId();
-            String contentIdString =
-                contentId != null ? contentId.toString() : contentIdFromContent(commitOp);
-            return KeyEntry.of(
-                buildIdentifiedKey(key, index, contentType, contentIdString, x -> null));
+            KeyEntry keyEntry =
+                KeyEntry.of(
+                    buildIdentifiedKey(key, index, contentType, contentIdString, x -> null));
+
+            if (withContentPredicate != null && withContentPredicate.test(keyEntry)) {
+              if (content == null) {
+                content =
+                    contentMapping.fetchContent(
+                        requireNonNull(commitOp.value(), "Required value pointer is null"));
+              }
+              return KeyEntry.of(buildIdentifiedKey(key, index, content, x -> null), content);
```

Comment:
`keyEntry.getKey()` could be reused instead of rebuilding it.

Reply:
done